# D599 Task 3 Program

A1. One question that is relevant to a real-world organizational situation that I will answer using market basket analysis is "Which products are frequently purchased together?" By identifying the products that are usually purchased together, it can help the organization create targeted discounts and recommend bundled offers to customers, resulting in higher sales.

A2. One goal of the market basket analysis is to enhance complementary sales for the organization. By identifying which products are often purchased together, the organization can develop targeted marketing campaigns to promote those product combinations, which leads to increased customer engagement and sales. 

B1. By definition, market basket analysis is "a data mining technique retailers use to increase sales by better understanding customer purchasing patterns" (Amruta, 2024, par. 1). The market basket technique analyzes the dataset by converting the original dataset into a transactions data. It groups the dataset by using the column OrderID, so that every unique OrderID represents a separate transaction in the dataset. Each transaction has a group of products that were purchased together. The market basket analysis focuses on identifying frequent itemsets. If there is a group of products that is purchased together frequently, it will be considered as a frequent itemset. Once frequent itemsets are identified, association rules are created. Association rules show the relationship between different items purchased together. There are a couple of expected outcomes when performing the market basket technique in the dataset. First, by discovering the products that are often purchased together, the organization can sell those products as a bundle at a discounted price, resulting in increased sales by encouraging customers to buy more items. Second, the organization can improve its product recommendations to customers.

B2. One example of a transaction in the dataset is the OrderID 581316. In this order, the customer purchased three products in a single transaction. The products purchased are: RED RETROSPOT SUGAR JAM BOWL, GLASS SONGBIRD STORAGE JAR, and REGENCY SUGAR BOWL GREEN. Looking at the dataset, while there are three separate rows for three different products, those three products still came from a single transaction because they have the same OrderID value and the same InvoiceDate value. This is an example of a single transaction that would be in the transaction data to be used for market basket analysis. 

B3. One assumption of market basket analysis is that the items frequently purchased together have an underlying relationship. These associations are not random and they reflect customer preferences and purchasing behavior, which can be used by organizations to create personalized recommendations or special promotions to customers. This results in higher sales for organizations and improves customer satisfaction.

In [30]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('C:/Users/jcaye/Downloads/Task 3 Dataset.csv')
df

OrderID                         ProductName  Quantity      InvoiceDate  \
0      536370         INFLATABLE POLITICAL GLOBE         48   12/1/2010 8:45   
1      536370      SET2 RED RETROSPOT TEA TOWELS         18   12/1/2010 8:45   
2      536370     PANDA AND BUNNIES STICKER SHEET        12   12/1/2010 8:45   
3      536370       RED TOADSTOOL LED NIGHT LIGHT        24   12/1/2010 8:45   
4      536370  VINTAGE HEADS AND TAILS CARD GAME         24   12/1/2010 8:45   
...       ...                                 ...       ...              ...   
8229   580263          DANISH ROSE DELUXE COASTER        12  12/2/2011 12:43   
8230   580263                CACTI TLIGHT CANDLES        16  12/2/2011 12:43   
8231   581316        RED RETROSPOT SUGAR JAM BOWL         1  12/8/2011 11:46   
8232   581316         GLASS  SONGBIRD STORAGE JAR         1  12/8/2011 11:46   
8233   581316            REGENCY SUGAR BOWL GREEN         1  12/8/2011 11:46   

     UnitPrice  TotalCost         Country DiscountApplied OrderPriority  \
0        $0.85     $40.80   United States             Yes          High   
1        $2.95     $53.10   United States             Yes          High   
2        $0.85     $10.20   United States             Yes          High   
3        $1.65     $39.60   United States             Yes          High   
4        $1.25     $30.00   United States             Yes          High   
...         ...        ...            ...             ...           ...   
8229     $0.85     $10.20   United States              No          High   
8230     $0.42      $6.72   United States              No          High   
8231     $2.55      $2.55   United States              No          High   
8232    $12.50     $12.50   United States              No          High   
8233     $4.15      $4.15   United States              No          High   

         Region    Segment ExpeditedShipping PaymentMethod  \
0     Northeast  Corporate               Yes   Credit Card   
1     Northeast  Corporate               Yes   Credit Card   
2     Northeast  Corporate               Yes   Credit Card   
3     Northeast  Corporate               Yes   Credit Card   
4     Northeast  Corporate               Yes   Credit Card   
...         ...        ...               ...           ...   
8229  Southeast  Corporate               Yes        PayPal   
8230  Southeast  Corporate               Yes        PayPal   
8231  Southeast   Consumer                No   Credit Card   
8232  Southeast   Consumer                No   Credit Card   
8233  Southeast   Consumer                No   Credit Card   

     CustomerOrderSatisfaction  
0                    Satisfied  
1                    Satisfied  
2                    Satisfied  
3                    Satisfied  
4                    Satisfied  
...                        ...  
8229              Dissatisfied  
8230              Dissatisfied  
8231            Very Satisfied  
8232            Very Satisfied  
8233            Very Satisfied  

[8234 rows x 14 columns]

This dataset is a sales transaction log for a retail business. Each row is an individual transaction that has orderid, productname, etc.|

In [31]:
df[df.duplicated()]

OrderID                          ProductName  Quantity  \
1686   547194     PLASTERS IN TIN WOODLAND ANIMALS         1   
1699   547194  SET OF 2 TEA TOWELS APPLE AND PEARS         1   
2440   552826                 HOT BATHS METAL SIGN         4   
5440   569332       SET2 RED RETROSPOT TEA TOWELS          1   
5445   569332   SET OF 2 TEA TOWELS PING MICROWAVE         1   
5449   569332   ROUND SNACK BOXES SET OF 4 FRUITS          1   
5459   569332               PLASTERS IN TIN SKULLS         1   
5465   569332                  SPACEBOY LUNCH BOX          1   
5485   569332               LUNCH BAG APPLE DESIGN         1   
6527   574506         ANTIQUE SILVER BAUBLE LAMP           3   
6688   575519         MINI PLAYING CARDS SPACEBOY         20   
6732   575661         ROCKING HORSE RED CHRISTMAS          3   
6791   575692             GLITTER HEART DECORATION         1   
7501   578541         RED APPLES CHOPPING BOARD            2   
7503   578541      VEGETABLE GARDEN CHOPPING BOARD         2   
8192   574095  ROUND SNACK BOXES SET OF4 WOODLAND          1   

           InvoiceDate UnitPrice  TotalCost         Country DiscountApplied  \
1686   3/21/2011 12:42     $1.65      $1.65   United States              No   
1699   3/21/2011 12:42     $2.95      $2.95   United States              No   
2440   5/11/2011 13:20     $2.10      $8.40   United States              No   
5440   10/3/2011 13:46     $3.25      $3.25   United States              No   
5445   10/3/2011 13:46     $2.95      $2.95   United States              No   
5449   10/3/2011 13:46     $2.95      $2.95   United States              No   
5459   10/3/2011 13:46     $1.65      $1.65   United States              No   
5465   10/3/2011 13:46     $1.95      $1.95   United States              No   
5485   10/3/2011 13:46     $1.65      $1.65   United States              No   
6527   11/4/2011 13:24    $10.40     $31.20   United States              No   
6688  11/10/2011 10:53     $0.42      $8.40   United States              No   
6732  11/10/2011 14:39     $0.85      $2.55   United States              No   
6791  11/10/2011 16:27     $0.39      $0.39   United States              No   
7501  11/24/2011 15:01     $4.95      $9.90   United States              No   
7503  11/24/2011 15:01     $4.95      $9.90   United States              No   
8192    11/3/2011 9:54     $2.95      $2.95   United States              No   

     OrderPriority     Region    Segment ExpeditedShipping PaymentMethod  \
1686          High  Northeast   Consumer                No        PayPal   
1699          High  Northeast   Consumer                No        PayPal   
2440        Medium  Southeast   Consumer               Yes   Credit Card   
5440          High  Northeast   Consumer               Yes   Credit Card   
5445          High  Northeast   Consumer               Yes   Credit Card   
5449          High  Northeast   Consumer               Yes   Credit Card   
5459          High  Northeast   Consumer               Yes   Credit Card   
5465          High  Northeast   Consumer               Yes   Credit Card   
5485          High  Northeast   Consumer               Yes   Credit Card   
6527        Medium  Northeast   Consumer               Yes   Credit Card   
6688        Medium  Northeast   Consumer               Yes   Credit Card   
6732        Medium  Southeast   Consumer               Yes   Credit Card   
6791          High  Northeast   Consumer                No   Credit Card   
7501          High  Northeast  Corporate               Yes        PayPal   
7503          High  Northeast  Corporate               Yes        PayPal   
8192        Medium  Northeast  Corporate               Yes   Credit Card   

     CustomerOrderSatisfaction  
1686         Very Dissatisfied  
1699         Very Dissatisfied  
2440            Very Satisfied  
5440            Very Satisfied  
5445            Very Satisfied  
5449            Very Satisfied  
5459            Very Satisfied  
5465            

In [32]:
df.drop_duplicates(inplace=True)

In [33]:
df[df.duplicated()]

Empty DataFrame
Columns: [OrderID, ProductName, Quantity, InvoiceDate, UnitPrice , TotalCost , Country, DiscountApplied, OrderPriority, Region, Segment, ExpeditedShipping, PaymentMethod, CustomerOrderSatisfaction]
Index: []

## 1st Part

### Identify 2 ordinal and 2 nominal variables

The two ordinal variables in the dataset that I chose are OrderPriority and CustomerOrderSatisfaction. The two nominal variables in the dataset that I chose are PaymentMethod and Region.

#### Ordinal variables

In [34]:
df['OrderPriority'].unique()

array(['High', 'Medium'], dtype=object)

The OrderPriority variable has two unique values: 'High' and 'Medium'. The OrderPriority variable is an ordinal variable because its values have a ranking, with 'High' being greater or more important than 'Medium'. The OrderPriority variable has ranked values but the differences between those values are not measurable in numerical terms.

In [35]:
df['CustomerOrderSatisfaction'].unique()

array(['Satisfied', 'Dissatisfied', 'Prefer not to answer',
       'Very Satisfied', 'Very Dissatisfied'], dtype=object)

The CustomerOrderSatisfaction variable has five unique values: 'Very Satisfied', 'Satisfied', 'Prefer not to answer', 'Dissatisfied', and 'Very Dissatisfied'. The CustomerOrderSatisfaction variable is an ordinal variable because its values have a clear and meaningful order where 'Very Satisfied' is better than 'Satisfied, and 'Satisfied' is better than 'Dissatisfied'. Also, since the differences between those values are not numerically measurable, it makes it ordinal.

#### Nominal variables

In [36]:
df['PaymentMethod'].unique()

array(['Credit Card', 'PayPal'], dtype=object)

The PaymentMethod variable has two unique values: 'Credit Card' and 'PayPal'. The PaymentMethod variable is a nominal variable because its values have no meaningful order and they are simply different labels. There is no value in the variable that is lower or higher than the other. They are simply different methods of payment. 

In [37]:
df['Region'].unique()

array(['Northeast', 'Southeast'], dtype=object)

The Region variable has two unique values: 'Northeast' and Southeast'. The Region variable is a nominal variable because its values have no ranking and they are simply labels for different areas. No region is greater than or less than the other. The values are just labels without any quantitative relatinships between them.

<pre>
ORDINAL Order Priority (High, Medium) - ordinal encoding (2, 1)
ORDINAL CustomerOrderSatisfaction (Very Satisfied, Satisfied, Prefer not to answer, Dissatisfied, Very Dissatisfied) - ordinal encoding (5, 4, 3, 2, 1)
NOMINAL PaymentMethod (Credit Card, Paypal) - label encoding (0 Credit Card, 1 PayPal)
NOMINAL Region (Northeast, Southeast) - one-hot encoding 
</pre>

### Perform encoding methods for each variable

In [38]:
df['OrderPriority'] = df['OrderPriority'].replace({'High': 2, 'Medium': 1})

I performed ordinal encoding on the OrderPriority variable because ordinal encoding is used for variables that have a hierarchy, which OrderPriority has in its values where 'High' is greater or more important than 'Medium'. Ordinal encoding converts the string values into integer values that reflect their natural order. In this code, I used the replace() function to encode 'High' as 2 and 'Medium' as 1 in the OrderPriority column. 'High' has a higher integer value since it is has a higher priority than 'Medium'. 

In [40]:
df['CustomerOrderSatisfaction'] = df['CustomerOrderSatisfaction'].replace({'Very Satisfied': 5, 'Satisfied': 4, 'Prefer not to answer': 3, 'Dissatisfied': 2, 'Very Dissatisfied': 1})

I performed ordinal encoding on the CustomerOrderSatisfaction variable because ordinal encoding is used for variables that have a meaningful order, which CustomerOrderSatisfaction has in its values where 'Very Satisfied' is better than 'Satisfied', and 'Satisfied' is better than 'Dissatisfied'. Ordinal encoding converts the values into integer values that reflect the order of the satisfaction levels. In this code, I used the replace() function to encode 'Very Satisfied' as 5, 'Satisfied' as 4, 'Prefer not to answer' as 3, 'Dissatisfied' as 2, and 'Very Dissatisfied' as 1 in the CustomerOrderSatisfaction column. 

In [41]:
df['PaymentMethod'] = df['PaymentMethod'].replace({'Credit Card': 0, 'PayPal': 1})

I performed label encoding on the PaymentMethod variable because label encoding is used for variables that have no hierarchy and have very few values, which describes PaymentMethod. Label encoding converts the string values into integer values and it will treat the values simply as different categories, not as ranked or ordered data. In this code, I used the replace() function to encode 'Credit Card' as 0 and 'PayPal' as 1 in the PaymentMethod column.

In [42]:
df = pd.get_dummies(df, columns=['Region'])

In [43]:
df['Region_Northeast'] = df['Region_Northeast'].astype(int)
df['Region_Southeast'] = df['Region_Southeast'].astype(int)

I performed one-hot encoding for the Region variable because one-hot encoding is appropriate for variables with values that have no meaningful order between them, which describes Region. One-hot encoding treats each value in the Region variable independently, which makes sense because there is no region that is less than or greater than the other. I used the get_dummies() function, which creates two new columns: Region_Northeast and Region_Southeast. Rows that had the value 'Northeast' in the Region column will have a value of 1 in the Region_Northeast column, 0 otherwise. Rows that had the value 'Southeast' in the Region column will have a value of 1 in the Region_Southeast column, 0 otherwise. I converted the two new columns into an int data type to have numeric values instead of boolean values. 

In [45]:
df

OrderID                         ProductName  Quantity      InvoiceDate  \
0      536370         INFLATABLE POLITICAL GLOBE         48   12/1/2010 8:45   
1      536370      SET2 RED RETROSPOT TEA TOWELS         18   12/1/2010 8:45   
2      536370     PANDA AND BUNNIES STICKER SHEET        12   12/1/2010 8:45   
3      536370       RED TOADSTOOL LED NIGHT LIGHT        24   12/1/2010 8:45   
4      536370  VINTAGE HEADS AND TAILS CARD GAME         24   12/1/2010 8:45   
...       ...                                 ...       ...              ...   
8229   580263          DANISH ROSE DELUXE COASTER        12  12/2/2011 12:43   
8230   580263                CACTI TLIGHT CANDLES        16  12/2/2011 12:43   
8231   581316        RED RETROSPOT SUGAR JAM BOWL         1  12/8/2011 11:46   
8232   581316         GLASS  SONGBIRD STORAGE JAR         1  12/8/2011 11:46   
8233   581316            REGENCY SUGAR BOWL GREEN         1  12/8/2011 11:46   

     UnitPrice  TotalCost         Country DiscountApplied  OrderPriority  \
0        $0.85     $40.80   United States             Yes              2   
1        $2.95     $53.10   United States             Yes              2   
2        $0.85     $10.20   United States             Yes              2   
3        $1.65     $39.60   United States             Yes              2   
4        $1.25     $30.00   United States             Yes              2   
...         ...        ...            ...             ...            ...   
8229     $0.85     $10.20   United States              No              2   
8230     $0.42      $6.72   United States              No              2   
8231     $2.55      $2.55   United States              No              2   
8232    $12.50     $12.50   United States              No              2   
8233     $4.15      $4.15   United States              No              2   

        Segment ExpeditedShipping  PaymentMethod  CustomerOrderSatisfaction  \
0     Corporate               Yes              0                          4   
1     Corporate               Yes              0                          4   
2     Corporate               Yes              0                          4   
3     Corporate               Yes              0                          4   
4     Corporate               Yes              0                          4   
...         ...               ...            ...                        ...   
8229  Corporate               Yes              1                          2   
8230  Corporate               Yes              1                          2   
8231   Consumer                No              0                          5   
8232   Consumer                No              0                          5   
8233   Consumer                No              0                          5   

      Region_Northeast  Region_Southeast  
0                    1                 0  
1                    1                 0  
2                    1                 0  
3                    1                 0  
4                    1                 0  
...                ...               ...  
8229                 0                 1  
8230                 0                 1  
8231                 0                 1  
8232                 0                 1  
8233                 0                 1  

[8218 rows x 15 columns]

After performing the encoding methods, this is the final dataframe. 

## 2nd Part

### Transactionalize the Data

Converting the original dataframe into a dataframe that can be used for market basket analsysis.

In [46]:
transactions = df.groupby('OrderID')['ProductName'].apply(list)
transactions

OrderID
536370    [INFLATABLE POLITICAL GLOBE , SET2 RED RETROSP...
536852    [POLKADOT RAIN HAT , VINTAGE HEADS AND TAILS C...
536974    [EDWARDIAN PARASOL RED, LUNCH BAG RED RETROSPO...
537065    [PARTY TIME PENCIL ERASERS, RED RETROSPOT PURS...
537463    [PINK POLKADOT CHILDRENS UMBRELLA, RED RETROSP...
                                ...                        
581001    [RETROSPOT CHILDRENS APRON, STRAWBERRY LUNCH B...
581171    [RED RETROSPOT CHILDRENS UMBRELLA, RED RETROSP...
581279    [CAKE STAND WHITE TWO TIER LACE, REGENCY CAKES...
581316    [RED RETROSPOT SUGAR JAM BOWL, GLASS  SONGBIRD...
581587    [BAKING SET 9 PIECE RETROSPOT , CHILDRENS APRO...
Name: ProductName, Length: 441, dtype: object

Dataframe with different orderID. Within each orderID, are the mutiple products bought within that transaction. 

In [47]:
import mlxtend
from mlxtend.preprocessing import TransactionEncoder

# Convert transactions into a format suitable for market basket analysis
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)

# Convert to DataFrame
df_transactions = pd.DataFrame(te_ary, columns=te.columns_)
df_transactions

50S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
0                            False               False   
1                            False               False   
2                            False               False   
3                            False               False   
4                            False               False   
..                             ...                 ...   
436                          False               False   
437                          False               False   
438                          False               False   
439                          False               False   
440                          False               False   

      I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
0                           False                     False   
1                           False                     False   
2                           False                     False   
3                           False                     False   
4                           False                     False   
..                            ...                       ...   
436                         False                     False   
437                         False                     False   
438                         False                     False   
439                         False                     False   
440                         False                     False   

      SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
0                               False                    False   
1                               False                    False   
2                               False                    False   
3                               False                    False   
4                               False                    False   
..                                ...                      ...   
436                             False                    False   
437                             False                    False   
438                             False                    False   
439                             False                    False   
440                             False                    False   

      TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
0                 False                   False                       False   
1                 False                   False                       False   
2                 False                   False                       False   
3                 False                   False                       False   
4                 False                   False                       False   
..                  ...                     ...                         ...   
436               False                   False                       False   
437               False                   False                       False   
438               False                   False                       False   
439               False                   False                       False   
440               False                   False                       False   

     12 EGG HOUSE PAINTED WOOD  ...  WRAP VINTAGE PETALS  DESIGN  \
0                        False  ...                        False   
1                        False  ...                        False   
2                        False  ...                        False   
3                        False  ...                        False   
4                        False  ...                        False   
..                         ...  ...                          ...   
436                      False  ...                        False   
437                      False  ...                        False   
438                      False  ...                        False   
439                      False  ...                        False   
440                      False  ...                        Fal

This is the dataset that has been converted into a binary matrix format to be used for market basket analysis. Each row represents a single transaction and each column represents a unique product. The values in the table are either True or False. True means the product was purchased within the transaction, and False otherwise. This data is now ready to be used by the Apriori algorithm to create association rules that show frequently purchased products.

Here, I converted the original dataset into a transaction data where market basket analysis can be performed. I started the process by grouping the products within each order using the groupby() function, and storing them into a variable called transactions. Then, I imported the TransactionEncoder() function from mlxtend because it is the function that converts the data into a format that is appropriate for market basket analysis. Using the transform() function, it converted all the data in the transactions variable into a binary matrix, where columns represent products and rows represent transactions. The binary matrix is now stored in the variable te_ary. The variable te_ary is then converted into a pandas dataframe where each column represents a product, and it can either have a value of True or False, depending whether the product was included in the transaction. Looking at the final dataframe, it has 441 transactions and 1,562 unique products.

<p>C1d. For C1a, the two ordinal variables that I chose are OrderPriority and CustomerOrderSatisfaction, and the two nominal variables that I chose are PaymentMethod and Region. The variable OrderPriority is an ordinal variable because its values 'High' and 'Medium' can be ranked where 'High' is greater or more important than 'Medium'. The variable CustomerOrderSatisfaction is also an ordinal variable because its values are ranked and have a meaningful order where 'Very Satisfied' is better than 'Satisfied', and 'Satisfied' is better than 'Dissatisfied'. The variable PaymentMethod is a nominal variable because its values 'Credit Card' and 'Paypal' have no meaningful order and they are simply different modes of payment. There is no payment that is better or worse than the other. The variable Region is also a nominal variable because its values 'Northeast' and 'Southeast' are not ranked and they are simply labels for different areas. </p>

<p>For C1b, I performed the appropriate encoding method for each variable. For the variable OrderPriority, I performed ordinal encoding because ordinal encoding is used for variables that have a hierarchy, which OrderPriority has in its values where 'High' is greater or more important than 'Medium'. Ordinal encoding converts the string values into integer values that reflect their natural order. I used the replace() function to encode 'High' as 2 and 'Medium' as 1 in the OrderPriority column. For the variable CustomerOrderSatisfaction, I performed ordinal encoding because ordinal encoding is used for variables that have a meaningful order, which CustomerOrderSatisfaction has in its values where 'Very Satisfied' is better than 'Satisfied', and 'Satisfied' is better than 'Dissatisfied'. Ordinal encoding converts the values into integer values that reflect the order of the satisfaction levels. I used the replace() function to encode 'Very Satisfied' as 5, 'Satisfied' as 4, 'Prefer not to answer' as 3, 'Dissatisfied' as 2, and 'Very Dissatisfied' as 1 in the CustomerOrderSatisfaction column. For the variable PaymentMethod, I performed label encoding because label encoding is used for variables that have no hierarchy and have very few values, which describes PaymentMethod. Label encoding converts the string values into integer values and it will treat the values simply as different categories, not as ranked or ordered data. I used the replace() function to encode 'Credit Card' as 0 and 'PayPal' as 1 in the PaymentMethod column. For the variable Region, I used one-hot encoding because one-hot encoding is appropriate for variables with values that have no meaningful order between them, which describes Region. One-hot encoding treats each value in the Region variable independently, which makes sense because there is no region that is less than or greater than the other. I used the get_dummies() function, which creates two new columns: Region_Northeast and Region_Southeast. Rows that had the value 'Northeast' in the Region column will have a value of 1 in the Region_Northeast column, 0 otherwise. Rows that had the value 'Southeast' in the Region column will have a value of 1 in the Region_Southeast column, 0 otherwise.</p>

<p>For C1c, I converted the original dataset into a transaction data where market basket analysis can be performed. I started the process by grouping the products within each order using the groupby() function, and storing them into a variable called transactions. Then, I imported the TransactionEncoder() function from mlxtend because it is the function that converts the data into a format that is appropriate for market basket analysis. Using the transform() function, it converted all the data in the transactions variable into a binary matrix, where columns represent products and rows represent transactions. The binary matrix is now stored in the variable te_ary. The variable te_ary is then converted into a pandas dataframe where each column represents a product, and it can either have a value of True or False, depending whether the product was included in the transaction. Looking at the final dataframe, it has 441 transactions and 1,562 unique products. The dataframe is now prepared and ready to be used by the Apriori algorithm to create association rules that show frequently purchased products.</p>

### Generate association rules

In [48]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# Apply the Apriori algorithm to find frequent itemsets
frequent_itemsets = apriori(df_transactions, min_support=0.01, use_colnames=True)

# Generate association rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

In [49]:
rules

antecedents  \
0                      ( DOLLY GIRL BEAKER)   
1         (CHARLOTTE BAG DOLLY GIRL DESIGN)   
2                      ( DOLLY GIRL BEAKER)   
3               (DOLLY GIRL CHILDRENS BOWL)   
4                      ( DOLLY GIRL BEAKER)   
...                                     ...   
85169        (PACK OF 6 SKULL PAPER PLATES)   
85170  (SET20 RED RETROSPOT PAPER NAPKINS )   
85171       (SET OF 9 BLACK SKULL BALLOONS)   
85172        (SET6 RED SPOTTY PAPER PLATES)   
85173          (PACK OF 6 SKULL PAPER CUPS)   

                                             consequents  antecedent support  \
0                      (CHARLOTTE BAG DOLLY GIRL DESIGN)            0.020408   
1                                   ( DOLLY GIRL BEAKER)            0.058957   
2                            (DOLLY GIRL CHILDRENS BOWL)            0.020408   
3                                   ( DOLLY GIRL BEAKER)            0.040816   
4                             (DOLLY GIRL CHILDRENS CUP)            0.020408   
...                                                  ...                 ...   
85169  (SET6 RED SPOTTY PAPER CUPS, SET OF 9 HEART SH...            0.049887   
85170  (SET6 RED SPOTTY PAPER CUPS, SET OF 9 HEART SH...            0.117914   
85171  (SET6 RED SPOTTY PAPER CUPS, SET OF 9 HEART SH...            0.058957   
85172  (SET6 RED SPOTTY PAPER CUPS, SET OF 9 HEART SH...            0.113379   
85173  (SET6 RED SPOTTY PAPER CUPS, SET OF 9 HEART SH...            0.056689   

       consequent support   support  confidence       lift  leverage  \
0                0.058957  0.011338    0.555556   9.423077  0.010135   
1                0.020408  0.011338    0.192308   9.423077  0.010135   
2                0.040816  0.015873    0.777778  19.055556  0.015040   
3                0.020408  0.015873    0.388889  19.055556  0.015040   
4                0.036281  0.013605    0.666667  18.375000  0.012865   
...                   ...       ...         ...        ...       ...   
85169            0.011338  0.011338    0.227273  20.045455  0.010772   
85170            0.011338  0.011338    0.096154   8.480769  0.010001   
85171            0.013605  0.011338    0.192308  14.134615  0.010536   
85172            0.011338  0.011338    0.100000   8.820000  0.010052   
85173            0.011338  0.011338    0.200000  17.640000  0.010695   

       conviction  zhangs_metric  
0        2.117347       0.912500  
1        1.212828       0.949880  
2        4.316327       0.967262  
3        1.602968       0.987842  
4        2.891156       0.965278  
...           ...            ...  
85169    1.279445       1.000000  
85170    1.093839       1.000000  
85171    1.221250       0.987470  
85172    1.098513       1.000000  
85173    1.235828       1.000000  

[85174 rows x 10 columns]

<p> C3. In this code, I used the Apriori Algorithm to generate frequent itemsets and association rules from the binary matrix data. I set the minimum support threshold to 1%, which means only itemsets with at least 1% support will be considered frequent. The frequent itemsets are stored in the variable frequent_itemsets. Then, I used the association_rules() function to generate association rules from the frequent itemsets. Lift is the metric that was used to generate the rules and this metric measures how likely multiple items are to be purchased together compared to being purchased independently. Rules that have a lift value of at least 1 are the only rules to be included in the association rules table because we are only interested in rules that have a positive association. The association rules will include support, lift, and confidence values.</p>

<p>C4. Here is the association rules table. This table describes relationships between different items that are often purchased together. Each association rule has support, confidence, and lift values. Support measures how often an item or an itemset appears in all of the transactions. Confidence measures how likely the consequent (right) will get purchased if the antecedent (left) is purchased. A higher confidence value means stronger association between the consequent and antecedent. Lift measures how likely multiple items are to be purchased together compared to being purchased independently. A rule that has a lift value of at least 1 indicates that the products are more likely to be purchased together than if they were purchased randomly.</p>

In [50]:
top_3_rules = rules.sort_values(by='lift', ascending=False).head(3)
top_3_rules

antecedents  \
82987  (CHILDRENS CUTLERY SPACEBOY , CARD DOLLY GIRL ...   
84821  (ALARM CLOCK BAKELIKE GREEN, PLASTERS IN TIN S...   
84811  (ALARM CLOCK BAKELIKE GREEN, PLASTERS IN TIN S...   

                                             consequents  antecedent support  \
82987  (CHILDRENS CUTLERY DOLLY GIRL , ROUND SNACK BO...            0.011338   
84821  (SET6 RED SPOTTY PAPER CUPS, ROUND SNACK BOXES...            0.011338   
84811  (SET6 RED SPOTTY PAPER CUPS, ALARM CLOCK BAKEL...            0.011338   

       consequent support   support  confidence  lift  leverage  conviction  \
82987            0.011338  0.011338         1.0  88.2  0.011209         inf   
84821            0.011338  0.011338         1.0  88.2  0.011209         inf   
84811            0.011338  0.011338         1.0  88.2  0.011209         inf   

       zhangs_metric  
82987            1.0  
84821            1.0  
84811            1.0

<p>C5. Here are the top three rules that have the highest lift values. The first rule has four products as antecedents and they are: 'ALARM CLOCK BAKELIKE PINK', 'ALARM CLOCK BAKELIKE RED' 'PLASTERS IN TIN CIRCUS PARADE' and 'SET6 RED SPOTTY PAPER PLATES'. As for the consequents, the rule has four of them and they are: 'ALARM CLOCK BAKELIKE GREEN', 'PLASTERS IN TIN SPACEBOY', 'ROUND SNACK BOXES SET OF4 WOODLAND' and 'SET6 RED SPOTTY PAPER CUPS'. The support value is 1.1%, which means the combination of antecedents and consequents in this rule appear together in 1.1% of the total transactions. The confidence value is 1.0, which means every time the antecedents are purchased, the consequents also get purchashed 100% of the time. The lift value is 88.2, which means both antecedents and consequents are 88.2 times more likely to be purchased together than if their purchases were independent.</p>
<p> The second rule has four products as antecedents and they are: 'ALARM CLOCK BAKELIKE RED', 'CHILDRENS CUTLERY SPACEBOY', 'ROUND SNACK BOXES SET OF4 WOODLAND' and 'SPACEBOY BIRTHDAY CARD'. As for the consequents, the rule has two of them and they are: 'ALARM CLOCK BAKELIKE PINK' and 'CARD DOLLY GIRL'. The support value is 1.1%, which means the combination of antecedents and consequents in this rule appear together in 1.1% of the total transactions. The confidence value is 1.0, which means every time the antecedents are purchased, the consequents also get purchased 100% of the time. The lift value is 88.2, which means both antecedents and consequents are 88.2 times more likely to be purchased together than if their purchases were independent.</p>
<p> The third rule has three products as antecedents and they are: 'ALARM CLOCK BAKELIKE RED', 'PLASTERS IN TIN CIRCUS PARADE', and 'SET6 RED SPOTTY PAPER PLATES'. As for the consequents, the rule has four of them and they are: 'ALARM CLOCK BAKELIKE PINK', 'PASTERS IN TIN SPACEBOY', 'ROUND SNACK BOXES SET OF4 WOODLAND', and 'SET6 RED SPOTTY PAPER CUPS'. The support value is 1.1%, which means the combination of antecedents and consequents in this rule appear together in 1.1% of the total transactions. The confidence value is 1.0, which means every time the antecedents are purchased, the consequents also get purchased 100% of the time. The lift value is 88.2, which means both antecedents and consequents are 88.2 times more likely to be purchased together than if their purchases were independent.
</p>

<p>Overall, these three rules represent a strong and reliable association between their antecedents and consequents, despite each rule only appearing in 1.1% of all transaction in the dataset. These rules can be used to create targeted discounts and recommend bundled offers to customers.</p>

<p>D1. Support, confidence, and lift values are important metrics in market basket analysis because these metrics help identify which combinations of products have strong relationships. Support measures how frequently an itemset appears in the dataset. A higher support value means an itemset appears frequently in transactions. Confidence measures the probability that a transaction containing the antecedent also contains the consequent. A higher confidence value means that when the antecedent is purchased, the consequent is also often purchased. Lift measures the strength of association between the antecedent and consequent compared to random co-occurrence. A higher lift value means the antecedent and consequent are purchased together more frequently than expected by chance.</p>

<p>Looking at the general association rules table, each rule has support, confidence, and lift values. These three metrics are essential in assessing the strength and relevance of the association rules. In the top 3 association rules table we created from earlier, we can see that all three rules are highly reliable because they have high confidence and lift values, which indicate that there is a high probability that the items get purchased together than independently.</p>


<p>D2. Conducting market basket analysis on the dataset holds practical significance. The association rules can be used to enhance product recommendations by recommending complementary products to customers while their shopping, which increases the likelihood of them buying more items. Using one of our association rules for example, knowing that 'DOLLY GIRL BEAKER' and 'DOLLY GIRL CHILDRENS BOWL' frequently get purchased together, if the customer buys 'DOLLY GIRL BEAKER' they should also get recommended to buy 'DOLLY GIRL CHILDRENS BOWL' because those two items frequently get purchased together. This benefits businesses because it could significantly increase their sales, while also increasing customer engagement by offering relevant suggestions. </p>
<p>Another way businesses can take advantage of the results from market basket analysis is to create targeted discounts and bundling. They can offer discounts on products that are frequently purchased together or bundle the products to entice customer to purchase multiple items, which increases overall sales. </p>

D3. To answer the question from part A1 about "Which products are frequently purchased together?", we have identified the products that were frequently purchased together based on the market basket analysis results. There are two recommended courses of actions for frequently purchased products. The first recommended action is to display frequently purchased together products during the online checkout process. For example, if a customer selects the item 'PLASTER IN TIN SPACEBOY', they should get recommended to buy 'ALARM CLOCK BAKELIKE PINK' as a complementary purchase because both items are often purchased together based on the market basket analysis results. This increases the likelihood of customers buying more items, and thus increases sales. The second recommended action is to bundle commonly purchased together products with a slight discount to encourage customers to buy multiple items together, increasing the overall order size. For example, bundling the frequently purchased together items like 'ALARM CLOCK BAKELIKE RED', 'CHILDRENS CUTLERY SPACEBOY' and 'SPACEBOY BIRTHDAY CARD' together with a slight discount, it encourages customers to purchase them all together. This drives more sales to the business.

### C2. Include a copy of the cleaned dataset

In [19]:
df.to_csv('cleaned_dataset.csv', index=False)

In [22]:
df

OrderID                         ProductName  Quantity      InvoiceDate  \
0      536370         INFLATABLE POLITICAL GLOBE         48   12/1/2010 8:45   
1      536370      SET2 RED RETROSPOT TEA TOWELS         18   12/1/2010 8:45   
2      536370     PANDA AND BUNNIES STICKER SHEET        12   12/1/2010 8:45   
3      536370       RED TOADSTOOL LED NIGHT LIGHT        24   12/1/2010 8:45   
4      536370  VINTAGE HEADS AND TAILS CARD GAME         24   12/1/2010 8:45   
...       ...                                 ...       ...              ...   
8229   580263          DANISH ROSE DELUXE COASTER        12  12/2/2011 12:43   
8230   580263                CACTI TLIGHT CANDLES        16  12/2/2011 12:43   
8231   581316        RED RETROSPOT SUGAR JAM BOWL         1  12/8/2011 11:46   
8232   581316         GLASS  SONGBIRD STORAGE JAR         1  12/8/2011 11:46   
8233   581316            REGENCY SUGAR BOWL GREEN         1  12/8/2011 11:46   

     UnitPrice  TotalCost         Country DiscountApplied  OrderPriority  \
0        $0.85     $40.80   United States             Yes              2   
1        $2.95     $53.10   United States             Yes              2   
2        $0.85     $10.20   United States             Yes              2   
3        $1.65     $39.60   United States             Yes              2   
4        $1.25     $30.00   United States             Yes              2   
...         ...        ...            ...             ...            ...   
8229     $0.85     $10.20   United States              No              2   
8230     $0.42      $6.72   United States              No              2   
8231     $2.55      $2.55   United States              No              2   
8232    $12.50     $12.50   United States              No              2   
8233     $4.15      $4.15   United States              No              2   

        Segment ExpeditedShipping  PaymentMethod  CustomerOrderSatisfaction  \
0     Corporate               Yes              0                          4   
1     Corporate               Yes              0                          4   
2     Corporate               Yes              0                          4   
3     Corporate               Yes              0                          4   
4     Corporate               Yes              0                          4   
...         ...               ...            ...                        ...   
8229  Corporate               Yes              1                          2   
8230  Corporate               Yes              1                          2   
8231   Consumer                No              0                          5   
8232   Consumer                No              0                          5   
8233   Consumer                No              0                          5   

      Region_Northeast  Region_Southeast  
0                    1                 0  
1                    1                 0  
2                    1                 0  
3                    1                 0  
4                    1                 0  
...                ...               ...  
8229                 0                 1  
8230                 0                 1  
8231                 0                 1  
8232                 0                 1  
8233                 0                 1  

[8218 rows x 15 columns]

This is the dataset that has been cleaned and undergone various encoding processes. Both OrderPriority and CustomerOrderSatisfaction variables were ordinally encoded, the PaymentMethod was label encoded, and the Region column was one-hot encoded. This cleaned data is now prepared for further data analysis.

In [ ]:
df_transactions.to_csv('cleaned_transactions_dataset.csv', index=False)

In [24]:
df_transactions

50S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
0                            False               False   
1                            False               False   
2                            False               False   
3                            False               False   
4                            False               False   
..                             ...                 ...   
436                          False               False   
437                          False               False   
438                          False               False   
439                          False               False   
440                          False               False   

      I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
0                           False                     False   
1                           False                     False   
2                           False                     False   
3                           False                     False   
4                           False                     False   
..                            ...                       ...   
436                         False                     False   
437                         False                     False   
438                         False                     False   
439                         False                     False   
440                         False                     False   

      SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
0                               False                    False   
1                               False                    False   
2                               False                    False   
3                               False                    False   
4                               False                    False   
..                                ...                      ...   
436                             False                    False   
437                             False                    False   
438                             False                    False   
439                             False                    False   
440                             False                    False   

      TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
0                 False                   False                       False   
1                 False                   False                       False   
2                 False                   False                       False   
3                 False                   False                       False   
4                 False                   False                       False   
..                  ...                     ...                         ...   
436               False                   False                       False   
437               False                   False                       False   
438               False                   False                       False   
439               False                   False                       False   
440               False                   False                       False   

     12 EGG HOUSE PAINTED WOOD  ...  WRAP VINTAGE PETALS  DESIGN  \
0                        False  ...                        False   
1                        False  ...                        False   
2                        False  ...                        False   
3                        False  ...                        False   
4                        False  ...                        False   
..                         ...  ...                          ...   
436                      False  ...                        False   
437                      False  ...                        False   
438                      False  ...                        False   
439                      False  ...                        False   
440                      False  ...                        Fal

This is the dataset that has been converted into a binary matrix format to be used for market basket analysis. Each row represents a single transaction, and each column represents a unique product. The values in the table are either True or False. True means the product was purchased within the transaction, and False otherwise. This data is now ready to be used by the Apriori algorithm to create association rules that show frequently purchased products. This dataset is named “cleaned_transactions_dataset.csv” and I have included it in my submission.

### References

Amruta. (2024, October 14). <i>Market basket analysis: A comprehensive guide for businesses</i>. Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/10/a-comprehensive-guide-on-market-basket-analysis/ 